### ISSUE TITLE
Node.appendChild only inserts the last child while inserting removed children

### ISSUE THREAD

**Issue by fatihpense** [xmldom/xmldom#343]
Hello maintainers! First of all, thanks for your efforts in maintaining a high-quality, standards-based library. I have been using it from time to time, and it is nice to know there is a dependable library that respects the details of XML.

**Bug**
When you remove children, put them into an array, and then append to the parent, it only adds the last child.

```javascript
  let input = `<root><A/><B/><C/></root>`;

  const doc = new DOMParser().parseFromString(input);

  let el = doc.documentElement;

  let arr = [];
  while (el.firstChild) {
    let node = el.removeChild(el.firstChild);

    arr.push(node);
  }

  for (let child1 of arr) {
    el.appendChild(child1);
    
    // if (el.firstChild) {
    //   el.insertBefore(child1, el.firstChild);
    // } else {
    //   el.appendChild(child1);
    // }
  }
  var s = new XMLSerializer();
  var xml = s.serializeToString(doc);
  console.log(xml);
```

This outputs `<root><C/></root>` when it should be `<root><A/><B/><C/></root>`.

**Workaround**
If you change inside the for loop:
```javascript
  for (let child1 of arr) {
    // el.appendChild(child1);

    if (el.firstChild) {
      el.insertBefore(child1, el.firstChild);
    } else {
      el.appendChild(child1);
    }
  }
```
Then it outputs correctly as `<root><C/><B/><A/></root>` (You can reverse the array before the loop if the order is important)

**Comment by karfau** [comment]
Looks quite serious, thx for the report.
I will look into it this week.
Marked it for 0.8.0 since I'm planning to release it for a while. If the fix is not so eaay, it could end up in 0.8.x.

**Comment by karfau** [comment]
So, finally debugged it to the point of being sure that I know what is going wrong.
Or at least the first part of it.
xmldom is currently not clearing the following attributes of the removed child node:
- `parentNode`
- `previousSibling`
- `nextSibling`

I also discovered that `childNodes` is not being cleaned up at the previous last index when removing a child, meaning after removing all three nodes, there `childNodes.length` is 0, but it has element `C` at the indexes 0, 1 and 2.

Of course, when a child that should be added still has a `parentNode` it needs to first be removed from the parent. Which is true for every node that the code in the description wants to add. Since `previousSibling` and `nextSibling` are still present on the previously dropped childNode, the removal happening as part of the first `appendChild` brings back `B` and `C` before adding `A` the order of the `childNodes` is now `B`, `C`, `A` but the `childNodes.length` is 1 and so on.

I will fix it as part of 0.8.0 since it's clearly a (serious) bug, that nobody would expect or should rely on.

the main culprit in this codebase is that whenever the spec says "this is a getter to x", in our case it's actually a reference.
